### PR TITLE
Preliminary F3DEX3 LVP support

### PR DESCRIPTION
**Pull request by aglab2**
This PR provides a preliminary support for [F3DEX3](https://github.com/HackerN64/F3DEX3/). Most of the work done here is chore to make sure F3DEX3 games boot while looking correctly with unaltered F3DEX2 display lists.

### CODE REVIEW

**Comment by weinerschnitzel** [comment]
Can't wait to see the sauraen's demos (and future F3DEX3 games) on GLideN64. Thank you so much for doing this!

**Comment by ariahiro64** [comment]
dev of the only released romhack with f3dex3 support it works reasonably well. 
image

File: src/GBI.cpp [hunk 1]
```diff
@@ -387,6 +396,43 @@ void GBIInfo::loadMicrocode(u32 uc_start, u32 uc_dstart, u16 uc_dsize)
...
+			current.negativeY = false;
+			current.fast3DPersp = false;
+			current.combineMatrices = false;
+
+			std::set<std::string> features;
+			{
+				// 0x180 is absolutely an overkill but it is ok for now
+				const char* name_end = (const char*)memchr(probe, ' ', 0x180);
+				size_t name_len = name_end - probe;
+				// It will look like F3DEX3_LVP_BrW_NOC
```
Comment (gonetz) [inline src/GBI.cpp:416]: // It will look like F3DEX3_LVP_BrZ_NOC

File: src/GBI.cpp [hunk 2]
```diff
@@ -387,6 +396,43 @@ void GBIInfo::loadMicrocode(u32 uc_start, u32 uc_dstart, u16 uc_dsize)
...
+				std::stringstream name_stream(name);
+				while (std::getline(name_stream, feature, '_'))
+				{
+					features.emplace(std::move(feature));
+				}
+			}
+
+			current.f3dex3.legacyVertexPipeline = features.find("LVP") != features.end();
+			current.f3dex3.noOcclusionPlane = features.find("NOC") != features.end();
+			current.f3dex3.branchOnZ = features.find("BrZ") != features.end();
```
Comment (gonetz) [inline src/GBI.cpp:428]: A nitpick: may be use case-insensitive compare there?

**Review by gonetz — CHANGES_REQUESTED** [review]
Well done!
I have just a few minor notices.

**Comment by gonetz** [comment]
> dev of the only released romhack with f3dex3 support it works reasonably well.

Where can I get that romhack?

**Comment by ariahiro64** [comment]
> > dev of the only released romhack with f3dex3 support it works reasonably well.
> 
> Where can I get that romhack?

an old as version was released for the oot hacking contest in hylian modding but if you want a build that isnt ancient i can dm it in discord. i just dont want to have a incomplete version leaking early (after final i dont care if every single build gets pushed to public)

**Comment by ariahiro64** [comment]
> > > dev of the only released romhack with f3dex3 support it works reasonably well.
> > 
> > 
> > Where can I get that romhack?
> 
> an old as version was released for the oot hacking contest in hylian modding but if you want a build that isnt ancient i can dm it in discord. i just dont want to have a incomplete version leaking early (after final i dont care if every single build gets pushed to public)

i meant i can give you a newer debug build

**Comment by aglab2** [comment]
Addressed various issues in commit https://github.com/gonetz/GLideN64/pull/2842/commits/fc9589b94686f67189c2b0d7fc52b83202d92629

**Review by gonetz — APPROVED** [review]
LGTM!

**Comment by gonetz** [comment]
> an old as version was released for the oot hacking contest in hylian modding but if you want a build that isnt ancient i can dm it in discord.

@ariahiro64, I sent you a message in discord.

**Pull request merged**